docs/source/extending/interval-example.rst

```diff
@@ -75,6 +75,19 @@ Function arguments and global values will thusly be recognized as belonging
 to ``interval_type`` whenever they are instances of ``Interval``.
 
 
+Type inference for Python annotations
+-------------------------------------
+
+While ``typeof`` is used to infer the Numba type of Python objects,
+``as_numba_type`` is used to infer the Numba type of Python types.  For simple
+cases, we can simply register that the Python type ``Interval`` corresponds with
+the Numba type ``interval_type``::
+
+   from numba.extending import as_numba_type
+
+   as_numba_type.register(Interval, interval_type)
```

Comment:
This would be simpler as `Interval._numba_type_ = interval_type`, which already informs `typeof` today.

However, this is only really possible in contrived cases. In all the extension types I've used so far, I've needed something like `IntervalType(self.dtype)` instead of just a global `IntervalType()`.

The mapping between python types and numba types is frequently one-to-many - so while `as_numba_type` is useful for the trivial cases you have being used here, it doesn't scale to all use-cases.

Reply:
I agree that the current way Numba requires explicit types up-front does not address all possible use cases in Python.  The `as_numba_type` registry is meant to work statically on type annotations at compile type, not dynamically on instances at runtime.  I believe the combination of a [generic type annotation](https://docs.python.org/3/library/typing.html#generics) and the `@as_numba_type.register` decorator does allow for something like the `LoggedVar` example at the link above.  That being said, I agree that there's still a lot of room to improve the flexibility of `jitclass`.  There are several open issues about so-called "generic" `jitclass`es.  That was something I was interested in once this PR is concluded.

Reply:
The comment:

> The as_numba_type registry is meant to work statically on type annotations at compile type, not dynamically on instances at runtime. 

needs to be documented.

Reply:
Done.